### PR TITLE
[Search] Move policyID into Search query syntax and remove policyIDs param

### PR DESCRIPTION
**Pull request by Kicu**
### Details
This PR moves storing `policyID` as a field of new searchQuery instead of separate param.

- Previously we stored currently selected `policyID` in `policyIDs` param on all search related pages, while it was stored as `policyID` on other pages.
   - `policyID` param stays unchanged, while `policyIDs` becomes `policyID` stored inside the query in all Search related pages
   - the new param is added to grammar and will be parsed and put on top level, which means we will be refering to it by `queryJSON.policyID` exactly the same as we do with `queryJSON.type` etc...
 - this PR might look complex because there are multiple places (and hacks) in navigation where we rewrite or copy `policyID` from params to inside search query and the other way around

_Note:_ ~as far as I can tell backend does not yet recognise `policyID` being sent as `query.policyID` and it does not filter results; this needs to be consulted with @luacmartins - probably some small change is required?~ Api expected to work and filter correctly.

### Fixed Issues

$ https://github.com/Expensify/App/issues/46592
PROPOSAL:

### Tests

 - go to `/search` and use workspace switcher to change policy
 - verify that ws got switched: network request to Search contains the policyId and returned results are filtered by policy
 - switch from Search to other views and see that workspace switcher behaves correctly,
 - test different cases of changing policyID and going back and forth between views

### QA Steps

 - go to `/search` and use workspace switcher to change policy
 - verify that workspace got switched and search results are filtered by policy ID
 - switch from Search to other views and see that workspace switcher behaves correctly
 - test different cases of changing policyID and going back and forth between views

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/45ab1335-a4c0-45e3-ada3-3494dede2b9e

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/51374ece-c148-4538-89ed-c376a55dc8e3

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c8b52e42-8fb4-43cd-af71-123b2c3bf0df

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/23618d52-bb91-47f2-b861-7adb13cb4642

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/98e33aef-8131-48c3-84f9-b79cb43e2010

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Review by WojtekBoman — APPROVED** [review]

**Comment by Kicu** [comment]
@rayane-djouah please note that - as is often the case - backend is not yet ready for policyID filtering I think :D because right now it is sent in a different part of payload.
I'm waiting for Carlos to confirm that everything is ok.

Another small issue that I noticed is that when clicking on status tabs (expenses | shared | drafts) the `policyID` will reset to `undefined`. 
That is deliberate for now, because clicking on these buttons will reset all currently chosen filters.
What I mean is these buttons simply point to: `type: Expense, status: all`, `type:expense, status: drafts` etc. regardless of any chosen advanced filters we always reset, so I made policyID behave the same because it is kinda like filter.
We can change of course.

**Comment by luacmartins** [comment]
Yea, I'll work on the policyID filter today.

File: src/hooks/useActiveWorkspaceFromNavigationState.ts [hunk 1]
```diff
@@ -12,22 +12,20 @@ import SCREENS from '@src/SCREENS';
  */
 function useActiveWorkspaceFromNavigationState() {
     // The last policyID value is always stored in the last route in BottomTabNavigator.
-    const activeWorkpsaceID = useNavigationState<BottomTabNavigatorParamList, string | undefined>((state) => {
+    const activeWorkspaceID = useNavigationState<BottomTabNavigatorParamList, string | undefined>((state) => {
```
Comment (luacmartins) [inline src/hooks/useActiveWorkspaceFromNavigationState.ts:15]: Nice catch

**Review by luacmartins — APPROVED** [review]
LGTM

**Comment by luacmartins** [comment]
>That is deliberate for now, because clicking on these buttons will reset all currently chosen filters.
What I mean is these buttons simply point to: type: Expense, status: all, type:expense, status: drafts etc. regardless of any chosen advanced filters we always reset query on clicking, so I made policyID behave the same way because it is kinda like another filter.

I agree with this for now. It feels a bit weird to have the field reset, but it is inline with the other filters too. So I agree we should keep it consistent.

**Comment by luacmartins** [comment]
API PR is merged, just waiting on deploy

**Comment by luacmartins** [comment]
API PR is merged. This should be working now.

**Comment by Kicu** [comment]
@rayane-djouah can you review please?

I tested with api (although I used staging api, not sure if deployed to prod yet) and I got back filtered results. I did have to clean my search snapshots from onyx but not sure why. In theory query including `policyID` should generate a unique hash and the results should be correctly visible.

Something to be double-checked.
Otherwise this PR should be good now 👍

**Comment by Kicu** [comment]
### filtering by policy
https://github.com/user-attachments/assets/b8485c99-f59d-4eb8-8251-71eaff885c90

**Review by rayane-d — CHANGES_REQUESTED** [review]
NAB

**Review by rayane-d — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



https://github.com/user-attachments/assets/8c28aa86-b13e-4448-97fa-0cc6dd1621bb


</details>

<details>
<summary>Android: mWeb Chrome</summary>



https://github.com/user-attachments/assets/6989b569-165b-41e4-abf0-2d2087e853a6


</details>

<details>
<summary>iOS: Native</summary>



https://github.com/user-attachments/assets/e1c80ca9-2dcb-44c0-852a-75c82f96a37c


</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/user-attachments/assets/479bce17-c6f7-441d-9a80-b9bad4732548


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/user-attachments/assets/2b774cdc-ecd8-4a89-8d0a-f8579f51a309

Screenshot 2024-08-26 at 7 51 02 PM

Screenshot 2024-08-26 at 7 51 24 PM

</details>

<details>
<summary>MacOS: Desktop</summary>



https://github.com/user-attachments/assets/3e740774-5c41-4879-ad72-63038059e19b


</details>

**Review by luacmartins — DISMISSED** [review]
LGTM. There's a backend bug when sending multiple policyIDs, I'm working on a fix. NAB for this PR though since users can only select a single policyID atm

**Comment by luacmartins** [comment]
@Kicu we have conflicts

File: src/libs/Navigation/linkTo/index.ts [hunk 2]
```diff
@@ -49,18 +49,18 @@ export default function linkTo(navigation: NavigationContainerRef<RootStackParam
...
+    const policyID = extractedPolicyID ?? policyIDFromState;
     const lastRoute = rootState?.routes?.at(-1);
 
     const isNarrowLayout = getIsNarrowLayout();
 
     const isFullScreenOnTop = lastRoute?.name === NAVIGATORS.FULL_SCREEN_NAVIGATOR;
 
-    // policyIDs is present only on SCREENS.SEARCH.CENTRAL_PANE and it's displayed in the url as a query param, on the other pages this parameter is called policyID and it's shown in the url in the format: /w/:policyID
-    if (policyID && !isFullScreenOnTop && !policyIDs) {
+    // Policy on SCREENS.SEARCH.CENTRAL_PANE can be present only as part of SearchQuery, while on other pages it's called `policyID` and stored in the url in the format: /w/:policyID/
```
[the remark this answers is not in the document]
Comment (Kicu) [reply src/libs/Navigation/linkTo/index.ts]: updated

**Review by luacmartins — APPROVED** [review]

**Pull request merged**